File: chapter 08/recipe 8.10/DniScQueuedComponent/DniScQCObj.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.EnterpriseServices;

namespace DniScQueuedComponent
{
	[InterfaceQueuing]
	public interface IQueuedComponent
	{
		void LogMessage(string message);
	}

	[ExceptionClass("DniScQueuedComponent.DniScQCErrorObj")]
	[ClassInterface(ClassInterfaceType.None)]
	public class DniScQCObj	
		: ServicedComponent, IQueuedComponent
	{
		public void LogMessage(string message)
		{
			//append the message to a file
			using (StreamWriter writer
				= new StreamWriter(@"c:\QCLog.txt", true))
			{
				writer.WriteLine(message);
				writer.Flush();
			}
		}
	}
}

File: chapter 08/recipe 8.10/DniScQueuedComponentVB/DniScQCObj.vb
Imports System.IO
Imports System.Runtime.InteropServices
Imports System.EnterpriseServices

<InterfaceQueuing()> _
Public Interface IQueuedComponent
    Sub LogMessage(ByVal message As String)
End Interface

<ExceptionClass("DniScQueuedComponentVB.DniScQCErrorObj")> _
<ClassInterface(ClassInterfaceType.None)> _
Public Class DniScQCObj
    Inherits ServicedComponent
    Implements IQueuedComponent

    Public Sub LogMessage(ByVal message As String) _
        Implements IQueuedComponent.LogMessage

        Dim writer As StreamWriter _
            = New StreamWriter("c:\QCLog.txt", True)
        writer.WriteLine(message)
        writer.Flush()
        writer.Close()
    End Sub
End Class
